### DISCUSSION TITLE
`hb_ot_color_glyph_reference_svg` returns blob of incorrect length

### DISCUSSION THREAD

**Discussion by krir17** [harfbuzz/harfbuzz#4945] [Q&A] [answered]
### Summary

The blob returned by [`hb_ot_color_glyph_reference_svg`](https://harfbuzz.github.io/harfbuzz-hb-ot-color.html#hb-ot-color-glyph-reference-svg) seems to contain the svg documents of all other glyphs that follow (instead of just the one) - with `hb_blob_get_data` returning a string that contains multiple `<svg>` tags : `<svg>...</svg><svg>...</svg>...</svg>`

### Investigation

Minimal test file, compiled with `g++ test.cpp $(pkg-config --cflags --libs harfbuzz) -o test.out` that takes a font file as input:

```cpp
#include <harfbuzz/hb.h>
#include <harfbuzz/hb-ot.h>
#include <iostream>

int main(int argc, char *argv[]) {
    const char* file = argv[1];
    hb_blob_t* blob = hb_blob_create_from_file(file);
    hb_face_t* face = hb_face_create(blob, 0);
    hb_blob_t* ref_svg_blob = hb_ot_color_glyph_reference_svg(face, 1406); // random glyphID for demonstration
    unsigned int length = hb_blob_get_length(ref_svg_blob);
    auto svg_string = hb_blob_get_data(ref_svg_blob, &length);
    std::cout << svg_string << std::endl;
}
```

Running `test.out` with `TwitterColorEmoji-SVGinOT.ttf` from https://github.com/13rac1/twemoji-color-font/releases/tag/v15.1.0 results in [out.txt](https://github.com/user-attachments/files/17822418/out.txt) - which contains svg documents for all glyphs starting from 1406.

I had a quick look at the code to see if this was intentional, and it looks like [`svgDocLength`](https://github.com/harfbuzz/harfbuzz/blob/b14def8bb32f32c32f2e2e9e1ce3efef2a242ca0/src/OT/Color/svg/svg.hh#L48) isn't initialized at this point. Perhaps, as a result, the value of `svgDocLength` is (almost) always larger than the 'parent' blob resulting in this weird behaviour?

### System Info
Harfbuzz version: 9.0.0
OS: Linux, Fedora 41

**Comment by khaledhosny** [comment, marked as answer]
`hb_blob_get_data()` does not return a null-terminated string, you need to use the returned length, e.g.: `std::cout.write(svg_string, length);`.

**Reply by krir17** [reply]
Ah that was what I was missing - thanks and sorry for the noise!
